File: backend/app/middleware/rate_limit.py
from __future__ import annotations

import asyncio
from collections.abc import Callable
from dataclasses import dataclass
from time import monotonic
from typing import Protocol


@dataclass(frozen=True, slots=True)
class RateLimitLease:
    allowed: bool
    limit: int
    remaining: int | None
    retry_after_seconds: float = 0.0


class RateLimiter(Protocol):
    async def acquire(self, bucket_id: str, rate_limit_per_sec: int) -> RateLimitLease:
        ...


@dataclass(slots=True)
class _BucketState:
    tokens: float
    updated_at: float


class InMemoryTokenBucketRateLimiter:
    def __init__(self, *, clock: Callable[[], float] = monotonic) -> None:
        self._clock = clock
        self._buckets: dict[str, _BucketState] = {}
        self._lock = asyncio.Lock()

    async def acquire(self, bucket_id: str, rate_limit_per_sec: int) -> RateLimitLease:
        if rate_limit_per_sec <= 0:
            return RateLimitLease(
                allowed=True,
                limit=rate_limit_per_sec,
                remaining=None,
            )

        async with self._lock:
            now = self._clock()
            refill_rate = float(rate_limit_per_sec)
            state = self._buckets.get(bucket_id)

            if state is None:
                state = _BucketState(tokens=refill_rate, updated_at=now)
                self._buckets[bucket_id] = state
            else:
                elapsed = max(now - state.updated_at, 0.0)
                state.tokens = min(refill_rate, state.tokens + (elapsed * refill_rate))
                state.updated_at = now

            if state.tokens >= 1.0:
                state.tokens -= 1.0
                return RateLimitLease(
                    allowed=True,
                    limit=rate_limit_per_sec,
                    remaining=max(int(state.tokens), 0),
                )

            retry_after_seconds = (1.0 - state.tokens) / refill_rate
            return RateLimitLease(
                allowed=False,
                limit=rate_limit_per_sec,
                remaining=0,
                retry_after_seconds=retry_after_seconds,
            )


_rate_limiter: RateLimiter | None = None


def get_rate_limiter() -> RateLimiter:
    global _rate_limiter

    if _rate_limiter is None:
        _rate_limiter = InMemoryTokenBucketRateLimiter()

    return _rate_limiter


def reset_rate_limiter() -> None:
    global _rate_limiter
    _rate_limiter = None

File: api/src/utils/rate-limit.ts
import type { RateLimitLease } from "../types";

interface BucketState {
  tokens: number;
  updatedAt: number;
}

export class InMemoryTokenBucketRateLimiter {
  private buckets = new Map<string, BucketState>();

  async acquire(bucketId: string, rateLimitPerSec: number): Promise<RateLimitLease> {
    if (rateLimitPerSec <= 0) {
      return {
        allowed: true,
        limit: rateLimitPerSec,
        remaining: null,
        retry_after_seconds: 0
      };
    }

    const now = Date.now() / 1000;
    const refillRate = rateLimitPerSec;
    const existing = this.buckets.get(bucketId);
    const state = existing ?? { tokens: refillRate, updatedAt: now };

    if (existing) {
      const elapsed = Math.max(now - state.updatedAt, 0);
      state.tokens = Math.min(refillRate, state.tokens + elapsed * refillRate);
      state.updatedAt = now;
    }

    if (state.tokens >= 1) {
      state.tokens -= 1;
      state.updatedAt = now;
      this.buckets.set(bucketId, state);
      return {
        allowed: true,
        limit: rateLimitPerSec,
        remaining: Math.max(Math.floor(state.tokens), 0),
        retry_after_seconds: 0
      };
    }

    this.buckets.set(bucketId, state);
    return {
      allowed: false,
      limit: rateLimitPerSec,
      remaining: 0,
      retry_after_seconds: (1 - state.tokens) / refillRate
    };
  }
}

let rateLimiter: InMemoryTokenBucketRateLimiter | null = null;

export function getRateLimiter(): InMemoryTokenBucketRateLimiter {
  if (rateLimiter == null) {
    rateLimiter = new InMemoryTokenBucketRateLimiter();
  }
  return rateLimiter;
}
